share/wake/lib/core/json.wake

```diff
@@ -538,3 +538,162 @@ export def lastValueInList (values: List JValue): Result JValue Error =
+        require (JInteger ji) = j else Nil
+        ji, Nil
+
+    view
+
+export def jeInteger: (Integer => Integer) => JValue => JValue =
+    def edit f j =
+        require (JInteger ji) = j else j
+        JInteger (f ji)
+
+    edit
+
+export def jBoolean: JValue => List Boolean =
+    def view j =
+        require (JBoolean jb) = j else Nil
+        jb, Nil
+
+    view
+
+export def jeBoolean: (Boolean => Boolean) => JValue => JValue =
+    def edit f j =
+        require (JBoolean jb) = j else j
+        JBoolean (f jb)
+
+    edit
+
+# Are these safe since 1.0 => JDouble 1.0 => to_json => from_json => JInteger 1?
+# Maybe jNumber is the only *safe* thing to use? On the other hand I'm worried
+# about only giving jNumber as an abstraction over jDouble...IDK yet.
+export def jDouble: JValue => List Double =
+    def view j =
+        require (JDouble jd) = j else Nil
+        jd, Nil
+
+    view
+
+export def jeDouble: (Double => Double) => JValue => JValue =
+    def edit f j =
+        require (JDouble jd) = j else j
+        JDouble (f jd)
+
+    edit
+
+export def jNumber: JValue => List Double =
+    def view = match _
+        JDouble x  = x, Nil
+        JInteger x = fromOption $ dint x
+        _          = Nil
+
+    view
+
+# This is odd because if you edit a JInteger with say +1 you still get a JDouble
+export def jeNumber: (Double => Double) => JValue => JValue =
+    def edit f s = match s
+        JDouble x  = JDouble (f x)
+        JInteger x =
+            require Some dx = dint x else s
+            JDouble (f dx)
+        _          = s
+
+    edit
+
+export def jNull: JValue => List Unit =
+    def view j =
+        require JNull = j else Nil
+        Unit, Nil
+
+    view
+
+export def jeNull: (Unit => Unit) => JValue => JValue =
+    def edit f j =
+        require JNull = j else j
+        def _ = f Unit # the edit function might be effectful
+        JNull
+
+    edit
+
+export def jObject: JValue => List (List (Pair String JValue)) =
+    def view j =
+        require (JObject jo) = j else Nil
+        jo, Nil
+
+    view
+
+export def jeObject: (List (Pair String JValue) => List (Pair String JValue)) => JValue => JValue =
+    def edit f j =
+        require (JObject jo) = j else j
+        JObject (f jo)
+
+    edit
+
+export def jArray: JValue => List (List JValue) =
+    def view j =
+        require (JArray ja) = j else Nil
+        ja, Nil
+
+    view
+
+export def jeArray: (List JValue => List JValue) => JValue => JValue =
+    def edit f j =
+        require (JArray jl) = j else j
+        JArray (f jl)
+
+    edit
+
+export def jSingelton: JValue => List JValue =
+    def view j =
+        require (JArray (j2, Nil)) = j else Nil
+        j2, Nil
+
+    view
+
+export def jeSingelton: (JValue => JValue) => JValue => JValue =
+    def edit f j =
+        require (JArray (j2, Nil)) = j else j
+        JArray (f j2, Nil)
+
+    edit
+
+export def jAll: JValue => List JValue = _ | jArray |> (_)
```

Comment:
... why bother with the two loops? `= mapFlat jArray`

Reply:
Or, for that matter, even more efficiently (no loops):
```
export def jAll: JValue => List JValue = match _
  JArray a = a
  _ = Nil
```

Reply:
Early example of how we have all the building blocks to compose things now and it also shows the symmetry with the jeAll case.

If we want to optimize for efficiency I think we can but otherwise I think this is good. It dog foods the query composition (but for sure its less efficient without some kind of fusion, no doubt)

Reply:
I think actually `_ | jArray | (_)` satisfies both of our complaints. It has no loops and does the same thing (the original only had one loop not two)

Reply:
`| (_)` is a no-op. I think you meant `|> (_)` ... which invokes `mapFlat` (a loop). I agree it's a loop over a single element, but any loop block wakes ability to inline the complexity away.

Reply:
Got it ok. We should probably make this as efficient as possible anyway, this is one of the most commonly used things (either directly or indirectly).